network/server.go

```diff
@@ -844,38 +973,45 @@ func (s *DefaultServer) SaveProtocolStream(
 		return
 	}
 
-	connectionInfo.addProtocolStream(protocol, stream)
+	connectionInfo.addProtocolClient(protocol, stream)
 }
 
 // NewProtoConnection opens up a new stream on the set protocol to the peer,
 // and returns a reference to the connection
 func (s *DefaultServer) NewProtoConnection(protocol string, peerID peer.ID) (*rawGrpc.ClientConn, error) {
-	s.protocolsLock.Lock()
-	defer s.protocolsLock.Unlock()
+	s.protocolsLock.RLock()
+	defer s.protocolsLock.RUnlock()
 
-	s.logger.Debug("NewProtoConnection", "protocol", protocol, "peer", peerID)
+	s.logger.Debug("new protocol connect", "protocol", protocol, "peer", peerID)
 
 	p, ok := s.protocols[protocol]
 	if !ok {
 		return nil, fmt.Errorf("protocol not found: %s", protocol)
 	}
 
+	s.logger.Debug("create new libp2p stream", "protocol", protocol, "peer", peerID)
+
 	stream, err := s.NewStream(protocol, peerID)
 	if err != nil {
 		return nil, err
 	}
 
-	// TODO: all connection use context background, need to be fixed
+	s.logger.Debug("create grpc client", "protocol", protocol, "peer", peerID)
+
+	// don't need context.WithTimeout, rpc client is fake
 	return p.Client(context.Background(), stream), nil
 }
 
 func (s *DefaultServer) NewStream(proto string, id peer.ID) (network.Stream, error) {
-	return s.host.NewStream(context.Background(), id, protocol.ID(proto))
+	// don't dial
+	ctx := network.WithNoDial(context.Background(), "grpc client stream")
```

Comment:
Could you add some more comments on why we should not dial here, and an additional comment on the method for any callers to notice?

Reply:
add it [f10b254](https://github.com/dogechain-lab/dogechain/pull/305/commits/f10b25407701e7017a7006b6c932c2b6967ffb20)